src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -180,85 +67,62 @@ module Read =
     let private resolvedEventLen (x: ITimelineEvent<EventBody>) = len x.Data + len x.Meta
     let private resolvedEventBytes events = events |> Array.sumBy resolvedEventLen
-    let private loggedReadSlice reader streamName batchSize requiresLeader startPos batchIndex (log: ILogger) ct: Task<_> = task {
+    let private loggedReadSlice reader streamName batchSize requiresLeader startPos batchIndex ct: Task<_> = task {
         let act = Activity.Current
         let! t, slice = readSliceAsync reader streamName batchSize startPos requiresLeader |> Stopwatch.time ct
         let bytes, count = slice.Messages |> resolvedEventBytes, slice.Messages.Length
-        let reqMetric: Log.Measurement = { stream = streamName; interval = t; bytes = bytes; count = count}
-        let evt = Log.Slice reqMetric
         if act <> null then act.IncMetric(count, bytes).AddLastVersion(slice.LastVersion) |> ignore
-        let log = if not (log.IsEnabled Events.LogEventLevel.Debug) then log else log |> Log.propResolvedEvents "Json" slice.Messages
-        (log |> Log.prop "startPos" startPos |> Log.prop "bytes" bytes |> Log.event evt).Information("Mdb{action:l} count={count} version={version}",
-            "Read", count, slice.LastVersion)
         return slice }
 
-    let private readBatches (log: ILogger) batchSize (readSlice: int64 -> int -> ILogger -> CancellationToken -> Task<StreamEventsSlice>)
+    let private readBatches fold initial tryDecode batchSize (readSlice: int64 -> int -> CancellationToken -> Task<StreamEventsSlice>)
             (maxPermittedBatchReads: int option) (startPosition: int64) ct
-        : Task<int64 * ITimelineEvent<EventBody>[]> =
+        : Task<int64 * 'state> =
         let mutable batchCount, pos = 0, startPosition
         let mutable version = -1L
-        let result = ResizeArray(int batchSize) // pre allocate batchSize as the vast majority of reads will only have a single batch
+        let mutable state = initial
         let rec loop () : Task<unit> = task {
             match maxPermittedBatchReads with
-            | Some mpbr when batchCount >= mpbr -> log.Information "batch Limit exceeded"; invalidOp "batch Limit exceeded"
+            | Some mpbr when batchCount >= mpbr -> invalidOp "batch Limit exceeded"
             | _ -> ()
 
-            let batchLog = log |> Log.prop "batchIndex" batchCount
-            let! slice = readSlice pos batchCount batchLog ct
+            let! slice = readSlice pos batchCount ct
             version <- max version slice.LastVersion
-            result.AddRange(slice.Messages)
+            state <- slice.Messages |> Array.chooseV tryDecode |> fold state
+            batchCount <- batchCount + 1
+            pos <- slice.LastVersion  + 1L
             if not slice.IsEnd then
-                batchCount <- batchCount + 1
-                pos <- slice.LastVersion  + 1L
                 return! loop () }
         task {
             do! loop ()
             let act = Activity.Current
```

Comment:
move this out to L125; as it does the top, it might as well do the tail too
at that point removing the loggedReadSlice layer in L70 becomes questionable - maybe it should also be inline into the consume loop (if backward loading ever became possible, the lack of DRYness can probably be handled by the Tracing.fs)